src/mlpack/methods/ann/layer/linear3d_impl.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file methods/ann/layer/linear3d_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Linear layer class which accepts 3D input.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "linear3d.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D() :
+    inSize(0),
+    outSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D(
+    const size_t inSize,
+    const size_t outSize,
+    RegularizerType regularizer) :
+    inSize(inSize),
+    outSize(outSize),
+    regularizer(regularizer)
+{
+  weights.set_size(outSize * inSize + outSize, 1);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Reset()
+{
+  typedef typename arma::Mat<typename OutputDataType::elem_type> MatType;
+
+  weight = MatType(weights.memptr(), outSize, inSize, false, false);
+  bias = MatType(weights.memptr() + weight.n_elem, outSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+template<typename eT>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  typedef typename arma::Mat<eT> MatType;
+  typedef typename arma::Cube<eT> CubeType;
+
+  if (input.n_rows % inSize != 0)
+  {
+    Log::Fatal << "Number of features in the input must be divisible by inSize."
+               << std::endl;
+  }
+
+  const size_t nPoints = input.n_rows / inSize;
+  const size_t batchSize = input.n_cols;
+
+  output.set_size(nPoints * outSize, batchSize);
+
+  CubeType inputTemp(const_cast<MatType&>(input).memptr(), nPoints, inSize,
+      batchSize, false, false);
```

Comment:
Do you need this particular ordering? Why didn't you use the following
```c++
CubeType inputTemp(const_cast<MatType&>(input).memptr(), inSize, nPoints,
    batchSize, false, false);
```
(I swapped `inSize` and `nPoints`).

Reply:
The reason is [similar to this one](https://github.com/mlpack/mlpack/pull/2398#issuecomment-667671075) but still, I'm wondering can we do something about it?? I mean suppose we switched to this version, then is it possible to get multihead attention working?

(Here `nPoints` : `seqLength`, `inSize` : `embeddingSize1` and `outSize` : `embeddingSize2`)
```cpp
arma::mat input(seqLength, batchSize);

// Following the style of keeping feature dimension to be the first. (This is not the case with recently merged Lookup layer)
// I'm assuming such a scenario.
FFN<LossType<>, InitType> model;
model.Add<Lookup<>>(vocabSize, embeddingSize1);  // The shape of output is : (embeddingSize1, seqLength, batchSize) if we keep feature dimension to be first.
model.Add<Linear3D<>>(embeddingSize1, embeddingSize2);   // The shape of output is : (embeddingSize2, seqLength, batchSize) if we keep feature dimension to be first.
model.Add<MultiheadAttention<>>(embeddingSize, numHeads);  // The shape of input will be (embeddingSize2, seqLength, batchSize), but we want the input of shape (seqLength, embeddingSize2, batchSize) to return an output of same shape.
```

Now if we follow the style of keeping feature dimension to be beside the batch dimension, the only problem is probably efficiency but it will be compatible with multihead-attention. What's your opinion?? I would definitely prefer efficiency but then how will it be compatible with multihead attention?
Let me know if something is unclear.

Reply:
> The reason is similar to this one but still, I'm wondering can we do something about it?? I mean suppose we switched to this version, then is it possible to get multihead attention working?

Perhaps, we could do the reordering before `MultiHeadAttention`. We could implement something like `TransposeLayer` which transposes each slice of the input. Although, the reordering will affect performance. Anyway, I can't say anything without benchmarks. So, let's leave it as is for now.

Reply:
>We could implement something like TransposeLayer.

That is actually a nice solution. But what if we do it inside the multihead attention layer? Should I try this thing?

Reply:
Don' waste your time. I'd like to try to take care of `dL/d(key)` and `dL/d(value)` first. Unfortunately, we have several PR's that depend on each other. If we change one, then we have to change the others.